### PR TITLE
Drop segment indexes when drop partition

### PR DESCRIPTION
**Pull request by xiaocai2333**
Signed-off-by: cai.zhang <cai.zhang@zilliz.com>
issue: #18642

### CODE REVIEW

**Comment by xiaocai2333** [comment]
rerun ut

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/indexcoord/index_coord.go [hunk 1]
```diff
@@ -635,15 +635,32 @@ func (i *IndexCoord) DropIndex(ctx context.Context, req *indexpb.DropIndexReques
 	for indexID := range indexID2CreateTs {
 		indexIDs = append(indexIDs, indexID)
 	}
-	err := i.metaTable.MarkIndexAsDeleted(req.CollectionID, indexIDs)
-	if err != nil {
-		ret.ErrorCode = commonpb.ErrorCode_UnexpectedError
-		ret.Reason = err.Error()
-		return ret, nil
+	if req.PartitionID == -1 {
```
Comment (czs007) [inline internal/indexcoord/index_coord.go:638]: do not use magic number.

File: internal/proxy/task.go [hunk 2]
```diff
@@ -1524,6 +1524,7 @@ func (dit *dropIndexTask) Execute(ctx context.Context) error {
 	var err error
 	dit.result, err = dit.indexCoord.DropIndex(ctx, &indexpb.DropIndexRequest{
 		CollectionID: dit.collectionID,
+		PartitionID:  -1,
```
Comment (czs007) [inline internal/proxy/task.go:1527]: do  not use magic number

File: internal/rootcoord/broker.go [hunk 3]
```diff
@@ -37,7 +37,7 @@ type Broker interface {
 	Flush(ctx context.Context, cID int64, segIDs []int64) error
 	Import(ctx context.Context, req *datapb.ImportTaskRequest) (*datapb.ImportTaskResponse, error)
 
-	DropCollectionIndex(ctx context.Context, collID UniqueID) error
+	DropCollectionIndex(ctx context.Context, collID UniqueID, partID UniqueID) error
```
Comment (czs007) [inline internal/rootcoord/broker.go:40]: partIDs []UniqueID

File: internal/rootcoord/step.go [hunk 4]
```diff
@@ -251,10 +251,12 @@ func (s *releaseCollectionStep) Weight() stepPriority {
 type dropIndexStep struct {
 	baseStep
 	collID UniqueID
+	// partID = -1 means to drop the index at the collection level
+	partID UniqueID
```
Comment (longjiquan) [inline internal/rootcoord/step.go:255]: `step` is just a wrapper of broker api. So just keep partIDs same with broker api here.

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: internal/rootcoord/drop_collection_task.go [hunk 5]
```diff
@@ -74,6 +74,7 @@ func (t *dropCollectionTask) Execute(ctx context.Context) error {
 	redoTask.AddAsyncStep(&dropIndexStep{
 		baseStep: baseStep{core: t.core},
 		collID:   collMeta.CollectionID,
+		partID:   -1,
```
Comment (longjiquan) [inline internal/rootcoord/drop_collection_task.go:77]: According to the gc logic of indexcoord, passing `-1` will lead indexcoord filter everything, so nothing will be collected. Just passing `nil` here.

File: internal/rootcoord/drop_partition_task.go [hunk 6]
```diff
@@ -68,6 +68,12 @@ func (t *dropPartitionTask) Execute(ctx context.Context) error {
 		ts:           t.GetTs(),
 	})
 
+	redoTask.AddAsyncStep(&dropIndexStep{
+		baseStep: baseStep{core: t.core},
+		collID:   t.collMeta.CollectionID,
+		partID:   partID,
```
Comment (longjiquan) [inline internal/rootcoord/drop_partition_task.go:74]: Use partIDs instead.

**Comment by longjiquan** [comment]
LGTM

**Pull request merged**